onnx/inliner/inliner.cc

```diff
@@ -246,103 +200,114 @@ class Specializer {
   }
 
   // Process a sub-graph, contained as an attribute in a control-flow op node.
-  void Transform(GraphProto& graph) {
+  // Since we need both pre-processing and post-processing in the traversal, we
+  // override the VisitGraph method.
+  void VisitGraph(GraphProto* graph) override {
     rename_scopes.emplace_back();
-    for (auto& x : *graph.mutable_input())
+    for (auto& x : *graph->mutable_input())
       Rename(*x.mutable_name());
-    for (auto& init : *graph.mutable_initializer())
+    for (auto& init : *graph->mutable_initializer())
       Rename(*init.mutable_name());
-    for (auto& y : *graph.mutable_output())
+    for (auto& y : *graph->mutable_output())
       Rename(*y.mutable_name());
-    for (auto& n : *graph.mutable_node())
-      Transform(n);
+    for (auto& n : *graph->mutable_node())
+      VisitNode(&n);
     rename_scopes.pop_back();
   }
 
  public:
-  // The main specialization method: specialize a FunctionProto for a particular call-site.
+  // Renames variables in a FunctionProto for inlining a particular call-site. This does the following:
+  // (i)  Rename all intermediate variables in the function to ensure that they are unique (wrt the main graph).
+  // (ii) Rename inputs and outputs using names of actual parameters.
   static void
-  Specialize(const NodeProto& callnode, FunctionProto& callee, std::string unique_suffix, NameGenerator& generator) {
-    AttributeMap map;
-    for (auto& attr : callnode.attribute()) {
-      map[attr.name()] = &attr;
-    }
-    auto lookup = [&](const std::string& name) -> const AttributeProto* {
-      auto iter = map.find(name);
-      return (iter != map.end()) ? iter->second : nullptr;
-    };
-    Specializer specializer(unique_suffix, generator, lookup);
+  Rename(const NodeProto& callnode, FunctionProto& callee, std::string unique_suffix, NameGenerator& generator) {
+    InliningRenamer renamer(unique_suffix, generator);
 
-    specializer.Bind<false>(*callee.mutable_input(), callnode.input());
-    specializer.Bind<true>(*callee.mutable_output(), callnode.output());
+    renamer.Bind<false>(*callee.mutable_input(), callnode.input());
+    renamer.Bind<true>(*callee.mutable_output(), callnode.output());
 
-    for (auto& n : *callee.mutable_node())
-      specializer.Transform(n);
+    renamer.VisitFunction(&callee);
   }
 };
 
 // Identify the set of all "input" variables used by a given node.
 // This includes the variables listed as node.input, as well as
 // implicit inputs referred to in any graph-valued-attribute of the node.
+// In the case of variables referenced in sub-graphs, only non-local variables
+// are treated as implicit inputs.
 
-struct ComputeUsedVars : public NodeVisitorBase {
-  std::vector<std::string> result;
+class ComputeInputs : private Visitor {
+ private:
+  std::vector<std::unordered_set<std::string>> namescopes;
+
+  bool InNestedScope() {
+    return !namescopes.empty();
+  }
+
+  std::unordered_set<std::string>& CurrentScope() {
```

Comment:
const?

Reply:
Not possible, since it returns a mutable set.